lib/parameter_substitution/expression.rb

```diff
@@ -37,6 +37,20 @@ def substitution_parameter_names
       @expression_list.map_compact(&:parameter_name)
     end
 
+    def method_names
+      @expression_list.map_compact do |expression|
```

Comment:
I think reduce is what you want to use here instead of the `map_compact.flatten` combination.

```
@expression_list.reduce([]) do |all_methods_used, expression|
  all_methods_used + methods_used_by_expression(expression)
end
```